### DISCUSSION TITLE
Validate atomWithStorage's deserialization

### DISCUSSION THREAD

**Discussion by PindaPixel** [pmndrs/jotai#2194] [Ideas] [answered]
Recently I stumbled upon a scenario where the interface of an object stored in localStorage changed because of a switch between different abstractions dealing with localStorage. It would be nice to be able to invalidate or possibly convert items in localStorage for users that still have something stored using a deprecated interface.

For example, consider this store:
```typescript
export type Todo = {
    content: string;
    done: boolean;
}

const todoStore = atomWithStorage<Todo[]>('to-do', []);

export const todoTotalCount = atom((get) => {
    return get(todoStore).length
});

export const todoDoneCount = atom((get) => {
    return get(todoStore).filter((item) => item.done).length;
});

export default todoStore;
```

When the item present in localStorage  is not an array, this application will crash because `filter` doesn't exist on `get(todoStore)`.

It would be nice to incorporate an assertion function such as:
```ts
function isTodo(item: unknown): item is Todo {
    return !!item && typeof item === 'object'
    && 'content' in item
    && typeof item.content === 'string'
    && 'done' in item
    && typeof item === 'boolean';
}

function validateStore(item: unknown): item is Todo[] {
    return Array.isArray(item) && item.every(isTodo);
}
```
And have it either use the default initial value or some sort of conversion function if the items are not of the correct type.

Perhaps a simple validate/convert key in the options object would suffice:
```ts
export function atomWithStorage<Value>(
  key: string,
  initialValue: Value,
  storage: AsyncStorage<Value>,
  unstable_options?: { 
    unstable_getOnInit?: boolean; 
    convert?: (item: unknown) => Value;
  }
): WritableAtom<
  Value | Promise<Value>,
  [SetStateActionWithReset<Value | Promise<Value>>],
  Promise<void>
>
```

Of course, alternatively one could use a new key for localStorage each time the interface changes.

**Comment by dai-shi** [comment, marked as answer]
That's exactly one of the purposes of `storage` option.
TBH, the default storage with JSON.parse isn't safe in production.
I think we should update the docs to clarify that...

**Reply by PindaPixel** [reply]
Oh I see, indeed I had no idea, it's even in the docs already: https://github.com/pmndrs/jotai/blob/main/docs/utilities/storage.mdx#L134 so that's my bad.
